File: Data Structure I/4. Stack & Queue/P20.cpp
// Date: Sun, Oct 3, 2021
// 20. Valid Parentheses (https://leetcode.com/problems/valid-parentheses/)

// time complexity: O(n)
// space complexity: O(n)
bool isValid(string s) {
	stack<char> mystack;
	for (char ch: s) {
		if (ch == '(' || ch == '{' || ch == '[')
			mystack.push(ch);
		else {
			if (mystack.empty())
				return false;
			else {
				char ch_top = mystack.top();
				if (ch == ')' && ch_top != '(' || ch == '}' && ch_top != '{' || ch == ']' && ch_top != '[')
					return false;
				mystack.pop();
			}
		} 
	}
	return mystack.empty();
}

// shorter version
bool isValid(string s) {
	stack<char> mystack;
	for (char& ch: s) {
		switch(ch) {
			case '(': mystack.push(')'); break;
			case '{': mystack.push('}'); break;
			case '[': mystack.push(']'); break;
			default:
				if (mystack.empty() || ch != mystack.top())
					return false;
				mystack.pop();
		}
	}
	return mystack.empty();
}

File: LeetCoding-Challenge/March-2022/P20.java
package leetcodingchallenge.march2022;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Date: March 13, 2022
 * 20. Valid Parentheses
 * https://leetcode.com/problems/valid-parentheses/
 */

public class P20 {
    // time: O(n)
    // space: O(1)
    public boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        for (char ch : s.toCharArray()) {
            switch (ch) {
                case '(':
                    stack.push(')');
                    break;
                case '[':
                    stack.push(']');
                    break;
                case '{':
                    stack.push('}');
                    break;
                default:
                    if (stack.isEmpty() || stack.pop() != ch) return false;
            }
        }
        return stack.isEmpty();
    }
}
